File: packages/java/src/main/java/org/realtimemail/StatePolicy.java
package org.realtimemail;

import java.time.Instant;

public final class StatePolicy {
  public TrustedDomainState evaluateDomainState(String domain, DomainStateSnapshot snapshot) {
    if (snapshot.revokedDomains().contains(domain)) return TrustedDomainState.REVOKED;
    if (snapshot.mutedDomains().contains(domain)) return TrustedDomainState.MUTED;
    return snapshot.trustedDomains().contains(domain) ? TrustedDomainState.TRUSTED : TrustedDomainState.REVOKED;
  }

  public MessageLifecycleState evaluateMessageState(RealtimeMailMessage message, MessageStateSnapshot snapshot) {
    if (snapshot.deletedMessageIds().contains(message.id())) return MessageLifecycleState.DELETED;
    if (snapshot.supersededMessageIds().contains(message.id())) return MessageLifecycleState.SUPERSEDED;
    var now = snapshot.now().orElseGet(Instant::now);
    if (message.expiresAt().isPresent() && !message.expiresAt().get().isAfter(now)) return MessageLifecycleState.EXPIRED;
    if (snapshot.dismissedMessageIds().contains(message.id())) return MessageLifecycleState.DISMISSED;
    return MessageLifecycleState.VISIBLE;
  }

  public boolean shouldDisplay(RealtimeMailMessage message, DomainStateSnapshot domainSnapshot, MessageStateSnapshot messageSnapshot) {
    return evaluateDomainState(message.domain(), domainSnapshot) == TrustedDomainState.TRUSTED
      && evaluateMessageState(message, messageSnapshot) == MessageLifecycleState.VISIBLE;
  }
}

File: packages/csharp/RealtimeMail/StatePolicy.cs
namespace RealtimeMail;

public sealed record DomainStateSnapshot(
    IReadOnlySet<string> TrustedDomains,
    IReadOnlySet<string> MutedDomains,
    IReadOnlySet<string> RevokedDomains
);

public sealed record MessageStateSnapshot(
    IReadOnlySet<string> DismissedMessageIds,
    IReadOnlySet<string> DeletedMessageIds,
    IReadOnlySet<string> SupersededMessageIds,
    DateTimeOffset? Now = null
);

public sealed class StatePolicy
{
    public TrustedDomainState EvaluateDomainState(string domain, DomainStateSnapshot snapshot)
    {
        if (snapshot.RevokedDomains.Contains(domain)) return TrustedDomainState.Revoked;
        if (snapshot.MutedDomains.Contains(domain)) return TrustedDomainState.Muted;
        return snapshot.TrustedDomains.Contains(domain) ? TrustedDomainState.Trusted : TrustedDomainState.Revoked;
    }

    public MessageLifecycleState EvaluateMessageState(RealtimeMailMessage message, MessageStateSnapshot snapshot)
    {
        if (snapshot.DeletedMessageIds.Contains(message.Id)) return MessageLifecycleState.Deleted;
        if (snapshot.SupersededMessageIds.Contains(message.Id)) return MessageLifecycleState.Superseded;
        if (message.ExpiresAt is not null && message.ExpiresAt <= (snapshot.Now ?? DateTimeOffset.UtcNow)) return MessageLifecycleState.Expired;
        if (snapshot.DismissedMessageIds.Contains(message.Id)) return MessageLifecycleState.Dismissed;
        return MessageLifecycleState.Visible;
    }

    public bool ShouldDisplay(RealtimeMailMessage message, DomainStateSnapshot domainSnapshot, MessageStateSnapshot messageSnapshot)
    {
        return EvaluateDomainState(message.Domain, domainSnapshot) == TrustedDomainState.Trusted
            && EvaluateMessageState(message, messageSnapshot) == MessageLifecycleState.Visible;
    }
}
